src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,255 @@
+
+## The story
+
+[Barbara] is contributing to an [OS](https://www.tockos.org) that supports
+running multiple applications on a single microcontroller. These
+microcontrollers have as little as 10's of kilobytes of RAM and 100's of
+kilobytes of flash memory for code. Barbara is writing a library that is used by
+multiple applications -- and is linked into each application -- so the library
+is very resource constrained. The library should support asynchronous operation,
+so that multiple APIs can be used in parallel within each (single-threaded)
+application.
+
+Barbara begins writing the library by trying to write a console interface, which
+allows byte sequences to be printed to the system console. Here is an example
+sequence of events for a console print:
+
+1. The interface gives the kernel a callback to call when the print finishes,
+   and gives the kernel the buffer to print.
+1. The kernel prints the buffer in the background while the app is free to do
+   other things.
+1. The print finishes.
+1. The app tells the kernel it is ready for the callback to be invoked, and the
+   kernel invokes the callback.
+
+Barbara tries to implement the API using
+[`core::future::Future`](https://doc.rust-lang.org/stable/core/future/trait.Future.html)
+so that the library can be compatible with the async Rust ecosystem. The OS
+kernel does not expose a Future-based interface, so Barbara has to implement
+`Future` by hand rather than using async/await syntax. She starts with a
+skeleton:
+
+```rust
+/// Passes `buffer` to the kernel, and prints it to the console. Returns a
+/// future that returns `buffer` when the print is complete. The caller must
+/// call kernel_ready_for_callbacks() when it is ready for the future to return. 
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    // TODO: Set the callback
+    // TODO: Tell the kernel to print `buffer`
+}
+
+struct PrintFuture;
+
+impl core::future::Future for PrintFuture {
+    type Output = &'static mut [u8];
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // TODO: Detect when the print is done, retrieve `buffer`, and return
+        // it.
+    }
+}
+```
+
+Note: All error handling is omitted to keep things understandable.
+
+Barbara begins to implement `print_buffer`:
+
+```rust
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    kernel_set_print_callback(callback);
+    kernel_start_print(buffer);
+    PrintFuture {}
+}
+
+// New! The callback the kernel calls.
+extern fn callback() {
+    // TODO: Wake up the currently-waiting PrintFuture.
+}
+```
+
+So far so good. Barbara then works on `poll`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            Poll::Ready(kernel_get_buffer_back())
+        }
+        Poll::Pending
```

Comment:
else or `return` missing here?

Reply:
Yes, there was supposed to be a `return`. Fixed.